core/src/nodes/handled_node.rs

```diff
@@ -0,0 +1,261 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use muxing::StreamMuxer;
+use nodes::node::{NodeEvent, NodeStream, Substream};
+use futures::prelude::*;
+use std::io::Error as IoError;
+use Multiaddr;
+
+/// Handler for the substreams of a node.
+///
+/// > Note: When implementing the various methods, don't forget that you have to register the
+/// > task that was the latest to poll and notify it.
+pub trait NodeHandler<TSubstream> {
+    /// Custom event that can be received from the outside.
+    type InEvent;
+    /// Custom event that can be produced by the handler and that will be returned by the swarm.
+    type OutEvent;
+    /// Information about a substream. Can be sent to the handler through a `NodeHandlerEndpoint`,
+    /// and will be passed back in `inject_substream` or `inject_outbound_closed`.
+    type OutboundOpenInfo;
+
+    /// Sends a new substream to the handler.
+    ///
+    /// The handler is responsible for upgrading the substream to whatever protocol it wants.
+    fn inject_substream(&mut self, substream: TSubstream, endpoint: NodeHandlerEndpoint<Self::OutboundOpenInfo>);
+
+    /// Indicates the handler that the inbound part of the muxer has been closed, and that
+    /// therefore no more inbound substream will be produced.
+    fn inject_inbound_closed(&mut self);
+
+    /// Indicates the handler that an outbound substream failed to open because the outbound
+    /// part of the muxer has been closed.
+    fn inject_outbound_closed(&mut self, user_data: Self::OutboundOpenInfo);
+
+    /// Indicates the handler that the multiaddr future has resolved.
+    fn inject_multiaddr(&mut self, multiaddr: Result<Multiaddr, IoError>);
+
+    /// Injects an event coming from the outside in the handler.
+    fn inject_event(&mut self, event: Self::InEvent);
```

Comment:
IIUC, these "inject_*" methods are callbacks. If so, why not use the more traditional "on" prefix?

Reply:
I consider the handler as a state machine. Calling these methods changes the state. They don't need to used as callbacks.